File: src/lib/db/models/question.ts
import { db } from '../client';
import { generateId } from '../../utils/auth';

export interface Question {
  id: string;
  topicId: string;
  text: string;
  options: string[];
  correctAnswer: number;
  explanation: string;
  createdAt: Date;
  updatedAt: Date;
}

export class QuestionModel {
  static async create(
    topicId: string,
    text: string,
    options: string[],
    correctAnswer: number,
    explanation: string
  ): Promise<Question> {
    const id = generateId();
    const timestamp = Math.floor(Date.now() / 1000);

    try {
      const result = await db.execute({
        sql: `INSERT INTO questions (id, topic_id, text, options, correct_answer, explanation, created_at, updated_at)
              VALUES (?, ?, ?, ?, ?, ?, ?, ?)
              RETURNING *`,
        args: [id, topicId, text, JSON.stringify(options), correctAnswer, explanation, timestamp, timestamp]
      });

      if (!result.rows?.[0]) {
        throw new Error('Failed to create question');
      }

      return this.mapQuestion(result.rows[0]);
    } catch (error) {
      console.error('Error creating question:', error);
      throw error;
    }
  }

  static async getByTopicId(topicId: string): Promise<Question[]> {
    try {
      const result = await db.execute({
        sql: 'SELECT * FROM questions WHERE topic_id = ? ORDER BY created_at ASC',
        args: [topicId]
      });

      return (result.rows || []).map(this.mapQuestion);
    } catch (error) {
      console.error('Error getting questions:', error);
      throw error;
    }
  }

  private static mapQuestion(row: any): Question {
    return {
      id: row.id,
      topicId: row.topic_id,
      text: row.text,
      options: JSON.parse(row.options),
      correctAnswer: row.correct_answer,
      explanation: row.explanation,
      createdAt: new Date(row.created_at * 1000),
      updatedAt: new Date(row.updated_at * 1000)
    };
  }
}

File: backend/src/lib/db/models/question.py
from datetime import datetime
import json
from typing import List, Optional
from backend.src.lib.db.client import dbClient
from backend.src.lib.utils.auth import generateId
from backend.src.lib.utils.dates import now, to_unix_timestamp, from_unix_timestamp
from shared.src.types import Question

class QuestionModel:
    @staticmethod
    async def create(
        topicId: str,
        text: str,
        options: List[str],
        correctAnswer: int,
        explanation: str
    ) -> Question:
        id = generateId()
        timestamp = to_unix_timestamp(now())

        try:
            result = await dbClient.execute(
                sql="""
                    INSERT INTO questions (id, topic_id, text, options, correct_answer, explanation, created_at, updated_at)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                    RETURNING *
                """,
                args=[id, topicId, text, json.dumps(options), correctAnswer, explanation, timestamp, timestamp]
            )

            if not result.rows or len(result.rows) == 0:
                raise Exception('Failed to create question')

            return QuestionModel.mapQuestion(result.rows[0])
        except Exception as e:
            print('Error creating question:', e)
            raise e

    @staticmethod
    async def getByTopicId(topicId: str) -> List[Question]:
        try:
            result = await dbClient.execute(
                sql='SELECT * FROM questions WHERE topic_id = ? ORDER BY created_at ASC',
                args=[topicId]
            )

            return [QuestionModel.mapQuestion(row) for row in (result.rows or [])]
        except Exception as e:
            print('Error getting questions:', e)
            raise e

    @staticmethod
    async def getById(id: str) -> Optional[Question]:
        try:
            result = await dbClient.execute(
                sql='SELECT * FROM questions WHERE id = ?',
                args=[id]
            )

            if not result.rows or len(result.rows) == 0:
                return None

            return QuestionModel.mapQuestion(result.rows[0])
        except Exception as e:
            print('Error getting question:', e)
            raise e

    @staticmethod
    async def update(
        id: str,
        text: str,
        options: List[str],
        correctAnswer: int,
        explanation: str
    ) -> Question:
        timestamp = to_unix_timestamp(now())

        try:
            result = await dbClient.execute(
                sql="""
                    UPDATE questions
                    SET text = ?, options = ?, correct_answer = ?, explanation = ?, updated_at = ?
                    WHERE id = ?
                    RETURNING *
                """,
                args=[text, json.dumps(options), correctAnswer, explanation, timestamp, id]
            )

            if not result.rows or len(result.rows) == 0:
                raise Exception('Question not found')

            return QuestionModel.mapQuestion(result.rows[0])
        except Exception as e:
            print('Error updating question:', e)
            raise e

    @staticmethod
    def mapQuestion(row: any) -> Question:
        try:
            return {
                'id': row.id,
                'topicId': row.topic_id,
                'text': row.text,
                'options': json.loads(row.options),
                'correctAnswer': row.correct_answer,
                'explanation': row.explanation,
                'createdAt': from_unix_timestamp(row.created_at),
                'updatedAt': from_unix_timestamp(row.updated_at)
            }
        except Exception as e:
            print('Error mapping question:', e)
            raise e
